From 553f37ab799636465fad2ef1e4377a1a4211de44 Mon Sep 17 00:00:00 2001
From: yuan junhui <yuanjunhui@eswincomputing.com>
Date: Tue, 30 Apr 2024 11:29:19 +0800
Subject: [PATCH 008/109] WIN2030-15279:fix:add feature

Changelogs:
01,add reset
02,modify powerOnOff ret msg
03,add power consum
04,add pvtinfo

Change-Id: I111d2845701bb382d3bfaeb714e4671e4a7674d1
---
 Core/Src/web-server.c | 197 +++++++++++++++++++++++++++++++++++++++---
 1 file changed, 183 insertions(+), 14 deletions(-)

diff --git a/Core/Src/web-server.c b/Core/Src/web-server.c
index f1dbd4b..2af848e 100644
--- a/Core/Src/web-server.c
+++ b/Core/Src/web-server.c
@@ -9475,16 +9475,24 @@ const unsigned char info_html[] ="<html lang=\"en\"> \
 												<button type=\"submit\">Exit</button> \
 											</form> \
                                         </div> \
-                                        <h2>Machine Status</h2> \
+                                       <h3>Machine Status</h3> \
 										<div class=\"power-status\"> \
 											Power On/Off:<button id=\"power-on-change\" >unknow</button>   <br> \
 										</div> \
-                                        <div class=\"basic-static\" > \
-                                            <h3>Basic Status</h3> \
+										<div class=\"reset\"> \
+											Reset:<button id=\"reset\" >Reset</button>   <br> \
+										</div> \
+										<div> \
+											<h3>Power Consumption</h3> \
+											Power Consumption:<input type=\"text\" id=\"power_consum\" value=\"0\" style=\"width: 60px;\" disabled><br> \
+											<button id=\"power-consum-refresh\">refresh</button> <br> \
+										</div> \
+                                        <div class=\"pvt-info\" > \
+                                            <h3>PVT info</h3> \
                                             CPU Temperature:<input type=\"text\" id=\"cpu_temp\" value=\"0\" style=\"width: 60px;\" disabled><br> \
                                             NPU Temperature:<input type=\"text\" id=\"npu_temp\" value=\"0\" style=\"width: 60px;\" disabled><br> \
                                             Fan Speed      :<input type=\"text\" id=\"fan_speed\" value=\"0\" style=\"width: 60px;\" disabled><br> \
-                                            <button id=\"basic-static-update\">update</button> <button id=\"basic-static-refresh\">refresh</button> \
+                                            <button id=\"pvt-info-refresh\">refresh</button> \
                                         </div> \
                                         <div class=\"net-work\" > \
                                             <h3>Network System</h3> \
@@ -9514,7 +9522,7 @@ const unsigned char info_html[] ="<html lang=\"en\"> \
                                                                 if(response.status===0){\n \
                                                                     alert(\"update success!\");\n \
                                                                 }else{\n \
-                                                                    alert(response.msg);\n \
+                                                                    alert(response.message);\n \
                                                                 }\n \
                                                                 // 请求成功时的回调函数\n \
                                                                 console.log('Network settings updated successfully.');\n \
@@ -9555,7 +9563,7 @@ const unsigned char info_html[] ="<html lang=\"en\"> \
                                                                 if(response.status===0){\n \
                                                                     alert(\"update success!\");\n \
                                                                 }else{\n \
-                                                                    alert(response.status);\n \
+                                                                    alert(response.message);\n \
                                                                 }\n \
                                                                 // 请求成功时的回调函数\n \
                                                                 console.log('power-on-change updated successfully.');\n \
@@ -9567,7 +9575,61 @@ const unsigned char info_html[] ="<html lang=\"en\"> \
                                                             }\n \
                                                         });\n\
                                                     });\n\
+													$('#reset').click(function() { \n \
+                                                        $.ajax({ \n\
+                                                            url: '/reset',\n \
+                                                            type: 'POST',\n \
+                                                            contentType: 'application/x-www-form-urlencoded', // 设置Content-Type \n \
+                                                            success: function(response) {\n \
+                                                                if(response.status===0){\n \
+                                                                    alert(\"reset success!\");\n \
+                                                                }else{\n \
+                                                                    alert(response.message);\n \
+                                                                }\n \
+                                                                // 请求成功时的回调函数\n \
+                                                                console.log('Reset successfully.');\n \
+                                                            },\n \
+                                                            error: function(xhr, status, error) {\n \
+                                                                alert(\"reset failed!\");\n \
+                                                                // 请求失败时的回调函数\n \
+                                                                console.error('Error Reset:', error);\n \
+                                                            }\n \
+                                                        });\n \
+                                                    });\n \
+													$('#power-consum-refresh').click(function() {\n \
+                                                        $.ajax({\n \
+                                                            url: '/power_consum',\n \
+                                                            type: 'GET',\n \
+                                                            success: function(response) {\n \
+                                                                $('#power_consum').val(response.data.power_consum);\n \
+                                                                console.log('power_consum refreshed successfully.');\n \
+                                                            },\n \
+                                                            error: function(xhr, status, error) {\n \
+                                                                // 请求失败时的回调函数\n \
+                                                                console.error('Error refreshing power_consum:', error);\n \
+                                                            }\n \
+                                                        });\n\
+                                                    });\n\
+													$('#pvt-info-refresh').click(function() {\n \
+                                                        $.ajax({\n \
+                                                            url: '/pvt_info',\n \
+                                                            type: 'GET',\n \
+                                                            success: function(response) {\n \
+                                                                $('#cpu_temp').val(response.data.cpu_temp);\n \
+																$('#npu_temp').val(response.data.npu_temp);\n \
+																$('#fan_speed').val(response.data.fan_speed);\n \
+                                                                console.log('pvt info refreshed successfully.');\n \
+                                                            },\n \
+                                                            error: function(xhr, status, error) {\n \
+                                                                // 请求失败时的回调函数\n \
+                                                                console.error('Error refreshing pvt info:', error);\n \
+                                                            }\n \
+                                                        });\n\
+                                                    });\n\
+													// --------------------after load page ------------- \n \
    													// $('#net-work-refresh').click(); \n\
+													$('#power-consum-refresh').click(); \n\
+													$('#pvt-info-refresh').click(); \n\
 													$.ajax({ \n \
 														url: '/power_status',\n \
 														type: 'GET',\n \
@@ -9630,7 +9692,6 @@ char* concatenate_strings(const char* str1, const char* str2) {
     // 初始化内存区域
     strcpy(new_str, str1);   // 复制第一个字符串
     strcat(new_str, str2);   // 拼接第二个字符串
-
     return new_str; // 返回新字符串
 }
 
@@ -9654,7 +9715,6 @@ void send_redirect(struct netconn *conn, const char *location,const char* cookie
                 "Content-Length: 0\r\n"
                 "Connection: close\r\n\r\n",
                 location);
-
     }
     printf("header: %d  %s \n",strlen(header),header);
     netconn_write(conn, header, strlen(header), NETCONN_COPY);
@@ -10006,15 +10066,47 @@ const char* process_header(const char *header) {
 // }
 static bool led_on = FALSE;
 
+
+
+
 int get_power_status(){
 	printf("TODO call get_power_status \n");
-	return 1;//poweron,TODO call
+	return 1;//1:powerON,0:powerOFF
 }
 int change_power_status(int status){//status 0:power off,1:power on
 	//pass
 	printf("TODO call change_power_status %d \n",status);
-	return 0;//0 success,TODO call
+	return 0;//0 success
+}
+
+int reset(){
+	printf("TODO call reset\n");
+	return 1;//0 success
 }
+
+int get_power_consum(){
+	printf("TODO call get_power_consum\n");
+	return 100;
+}
+
+typedef struct {
+    int cpu_temp;
+    int npu_temp;
+    int fan_speed;
+} PVTInfo;
+
+
+
+PVTInfo get_pvt_info() {
+	printf("TODO call get_pvt_info\n");
+    PVTInfo pvtInfo;
+    pvtInfo.cpu_temp = 1;
+    pvtInfo.npu_temp = 2;
+    pvtInfo.fan_speed = 3;
+    return pvtInfo;
+}
+
+
  /** Serve one HTTP connection accepted in the http thread */
  static void
  http_server_netconn_serve(struct netconn *conn)
@@ -10207,6 +10299,50 @@ int change_power_status(int status){//status 0:power off,1:power on
 
                 netconn_write(conn, response_header, strlen(response_header), NETCONN_COPY);
                 netconn_write(conn, json_response, strlen(json_response), NETCONN_COPY);
+			}else if(strcmp(path, "/power_consum")==0 ){
+				int power_consum=get_power_consum();
+				char json_response[BUF_SIZE_128]={0};
+				 // 创建JSON格式的字符串
+                char *json_response_patt = "{\"status\":0,\"message\":\"success\",\"data\":{\"power_consum\":\"%d\"}}";//0 success, msg
+                sprintf(json_response, json_response_patt,power_consum);
+
+                // 发送HTTP头部
+                const char *header = "HTTP/1.1 200 OK\r\n"
+                                    "Content-Type: application/json\r\n"
+                                    "Connection: close\r\n"
+                                    "Content-Length: %d\r\n\r\n";
+
+                char response_header[BUF_SIZE_256];
+                sprintf(response_header, header, strlen(json_response));
+
+                printf("strlen(response_header):%d,strlen(json_response):%d \n",strlen(response_header),strlen(json_response));
+
+                netconn_write(conn, response_header, strlen(response_header), NETCONN_COPY);
+                netconn_write(conn, json_response, strlen(json_response), NETCONN_COPY);
+				
+			}else if(strcmp(path, "/pvt_info")==0 ){
+				printf("GET location: pvt_info \n");
+				PVTInfo pvt_info=get_pvt_info();
+
+				char json_response[BUF_SIZE_128]={0};
+				 // 创建JSON格式的字符串
+                char *json_response_patt = "{\"status\":0,\"message\":\"success\",\"data\":{\"cpu_temp\":\"%d\",\"npu_temp\":\"%d\",\"fan_speed\":\"%d\"}}";
+                sprintf(json_response, json_response_patt,pvt_info.cpu_temp,pvt_info.npu_temp,pvt_info.fan_speed);
+
+                // 发送HTTP头部
+                const char *header = "HTTP/1.1 200 OK\r\n"
+                                    "Content-Type: application/json\r\n"
+                                    "Connection: close\r\n"
+                                    "Content-Length: %d\r\n\r\n";
+
+                char response_header[BUF_SIZE_256];
+                sprintf(response_header, header, strlen(json_response));
+
+                printf("strlen(response_header):%d,strlen(json_response):%d \n",strlen(response_header),strlen(json_response));
+
+                netconn_write(conn, response_header, strlen(response_header), NETCONN_COPY);
+                netconn_write(conn, json_response, strlen(json_response), NETCONN_COPY);
+
 			}else if(strcmp(path, "/network")==0 ){
 				printf("GET location: network \n");
              
@@ -10508,8 +10644,7 @@ int change_power_status(int status){//status 0:power off,1:power on
                     send_redirect(conn,"/login.html",resp_cookies);
 
                 }else if(strcmp(path, "/power_status")==0 ){
-					
-					
+					printf("POST ,location: power_status \n");
                     char* status=NULL;
                     assert(p_params!=NULL);
                     kv_pair *current = params.head;
@@ -10531,9 +10666,16 @@ int change_power_status(int status){//status 0:power off,1:power on
 
 				
 					// 创建JSON格式的字符串
-                    char *json_response_patt = "{\"status\":%d,\"message\":\"success!\",\"data\":{}}";//0 success, msg
+                    char *json_response_patt =NULL; 
 					char json_response[BUF_SIZE_64]={0};
-					sprintf(json_response, json_response_patt, change_status_ret);
+
+					if(change_status_ret==0){
+						json_response_patt="{\"status\":%d,\"message\":\"success!\",\"data\":{}}";//0 success, msg
+						sprintf(json_response, json_response_patt, change_status_ret);
+					}else{
+						json_response_patt="{\"status\":%d,\"message\":\"error retcode %d\",\"data\":{}}";//0 success, msg
+						sprintf(json_response, json_response_patt, change_status_ret,change_status_ret);
+					}
 
                     // 发送HTTP头部
                     const char *header = "HTTP/1.1 200 OK\r\n"
@@ -10548,6 +10690,33 @@ int change_power_status(int status){//status 0:power off,1:power on
                     netconn_write(conn, response_header, strlen(response_header), NETCONN_COPY);
                     netconn_write(conn, json_response, strlen(json_response), NETCONN_COPY);
 
+				}else if(strcmp(path, "/reset")==0 ){
+					int reset_ret=reset();
+					// 创建JSON格式的字符串
+                    char *json_response_patt =NULL; 
+					char json_response[BUF_SIZE_64]={0};
+					if(reset_ret==0){
+						json_response_patt="{\"status\":%d,\"message\":\"success!\",\"data\":{}}";//0 success, msg
+						sprintf(json_response, json_response_patt, reset_ret);
+					}else{
+						json_response_patt="{\"status\":%d,\"message\":\"error retcode %d\",\"data\":{}}";//0 success, msg
+						sprintf(json_response, json_response_patt, reset_ret,reset_ret);
+					}
+
+                    // 发送HTTP头部
+                    const char *header = "HTTP/1.1 200 OK\r\n"
+                                        "Content-Type: application/json\r\n"
+                                        "Connection: close\r\n"
+                                        "Content-Length: %d\r\n\r\n";
+                    char response_header[256];
+                    sprintf(response_header, header, strlen(json_response));
+
+                    printf("strlen(response_header):%d,strlen(json_response):%d \n",strlen(response_header),strlen(json_response));
+
+                    netconn_write(conn, response_header, strlen(response_header), NETCONN_COPY);
+                    netconn_write(conn, json_response, strlen(json_response), NETCONN_COPY);
+
+
 				}else if(strcmp(path, "/network")==0 ){		//post network
                     printf("POST ,location: network \n");
 
-- 
2.25.1

